src/zenml/integrations/gcp_secrets_manager/secrets_manager/gcp_secrets_manager.py

```diff
@@ -0,0 +1,260 @@
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+import re
+from typing import Any, ClassVar, Dict, List
+
+from google.cloud import secretmanager
+
+from zenml.logger import get_logger
+from zenml.secret.base_secret import BaseSecretSchema
+from zenml.secret.secret_schema_class_registry import SecretSchemaClassRegistry
+from zenml.secrets_managers.base_secrets_manager import BaseSecretsManager
+from zenml.stack.stack_component_class_registry import (
+    register_stack_component_class,
+)
+
+logger = get_logger(__name__)
+
+ZENML_SCHEMA_NAME = "zenml-schema-name"
+ZENML_GROUP_KEY = "zenml-group-key"
+
+
+def add_group_name_to_keys(secret: BaseSecretSchema) -> Dict:
+    """This function adds the secret group name to the keys of each
+    secret key-value pair to allow using the same key across multiple
+    secrets.
+
+    Args:
+        secret: The ZenML Secret schema
+    """
+    new_dict = dict()
+    for k, v in secret.content.items():
+        new_dict[secret.name + "_" + k] = v
+
+    return new_dict
+
+
+def remove_group_name_from_key(combined_key_name: str, group_name: str) -> str:
+    """This function serves to remove the secret group name from the secret
+    key.
+
+    Args:
+        combined_key_name: Full name as it is within the gcp secrets manager
+        group_name: Group name (the ZenML Secret name)
+    Returns:
+        The cleaned key
+    """
+    if combined_key_name.startswith(group_name + '_'):
+        return combined_key_name[len(group_name + '_'):]
+    else:
+        raise RuntimeError(f"Key-name `{combined_key_name}` does not have the "
+                           f"prefix `{group_name}`. Key could not be "
+                           f"extracted.")
+
+
+@register_stack_component_class
+class GCPSecretsManager(BaseSecretsManager):
+    """Class to interact with the AWS secrets manager."""
+
+    project_id: str
+
+    # Class configuration
+    FLAVOR: ClassVar[str] = "gcp_secrets_manager"
+    CLIENT: ClassVar[Any] = None
+
+    @classmethod
+    def _ensure_client_connected(cls) -> None:
+        if cls.CLIENT is None:
+            cls.CLIENT = secretmanager.SecretManagerServiceClient()
+
+    @property
+    def parent_name(self):
+        return f"projects/{self.project_id}"
+
+    def register_secret(self, secret: BaseSecretSchema) -> None:
+        """Registers a new secret.
+
+        Args:
+            secret: the secret to register"""
+        self._ensure_client_connected()
+
+        if secret.name in self.get_all_secret_keys():
+            raise KeyError(
```

Comment:
A little nitpicky, but a `KeyError` is usually when something doesn't exist, not for duplicates

Reply:
What Exception would you suggest, I disliked every other Exception more

Reply:
Yep that's the shitty part, there isn't any 😄 
In `zenml.exceptions` we already have similar ones: `StackExistsError`, `StackComponentExistsError`, ...
Not sure if we should make a general one, or create new ones like `SecretExistsError` for these single-use cases